### PR TITLE
test: p2p: check misbehavior for non-continuous headers messages

### PR DESCRIPTION
**Pull request by theStack**
This PR adds missing test coverage for a peer sending a `headers` message where the headers don't connect to each other, which should be treated as misbehaving (not disconnecting though, as the score increase is only 20). The relevant code path is `PeerManagerImpl::ProcessHeadersMessage` -> `PeerManagerImpl::CheckHeadersPoW` -> `PeerManagerImpl::CheckHeadersAreContinuous`:

https://github.com/bitcoin/bitcoin/blob/17acb2782a66f033238340e4fb81009e7f79e97a/src/net_processing.cpp#L2415-L2419

https://github.com/bitcoin/bitcoin/blob/17acb2782a66f033238340e4fb81009e7f79e97a/src/net_processing.cpp#L2474-L2484

### CODE REVIEW

**Comment by theStack** [comment]
(Force-pushed with a simpler variant that just deletes a random block header in the middle of a valid `headers` message instead of messing around with `hashPrevBlock`, which needed to re-grind the header for valid PoW. The previous version looked like this:
```
        # modify previous block hash of an arbitrary sequent block header to break link
        modified_block_header = block_headers[random.randrange(1, len(block_headers))]
        modified_block_header.hashPrevBlock ^= 1
        # need to re-calculate for valid PoW, as this check is done first
        modified_block_header.rehash()
        while not modified_block_header.hash.startswith('0'):
            modified_block_header.nNonce += 1
            modified_block_header.rehash()
```

)

File: test/functional/p2p_invalid_messages.py [hunk 1]
```diff
@@ -283,6 +285,25 @@ def test_invalid_pow_headers_msg(self):
             peer.send_message(msg_headers([blockheader]))
             peer.wait_for_disconnect()
 
+    def test_noncontinuous_headers_msg(self):
+        self.log.info("Test headers message with non-continuous headers sequence is logged as misbehaving")
+        block_hashes = self.generate(self.nodes[0], 10)
+        block_headers = []
```
Comment (instagibbs) [inline test/functional/p2p_invalid_messages.py:291]: absolutely don't have to take, I just like using this pythonism

```suggestion
        block_headers = [from_hex(CBlockHeader(), self.nodes[0].getblockheader(block_hash, False) for block_hash in block_hashes]
```

File: test/functional/p2p_invalid_messages.py [hunk 2]
```diff
@@ -283,6 +285,25 @@ def test_invalid_pow_headers_msg(self):
...
+        MISBEHAVING_NONCONTINUOUS_HEADERS_MSGS = ['Misbehaving', 'non-continuous headers sequence']
+        peer = self.nodes[0].add_p2p_connection(P2PInterface())
+        with self.nodes[0].assert_debug_log([], unexpected_msgs=MISBEHAVING_NONCONTINUOUS_HEADERS_MSGS):
+            peer.send_and_ping(msg_headers(block_headers))
+
+        # delete arbitrary block header somewhere in the middle to break link
+        del block_headers[random.randrange(1, len(block_headers)-1)]
+        with self.nodes[0].assert_debug_log(expected_msgs=MISBEHAVING_NONCONTINUOUS_HEADERS_MSGS):
+            peer.send_and_ping(msg_headers(block_headers))
+        self.nodes[0].disconnect_p2ps()
```
Comment (instagibbs) [inline test/functional/p2p_invalid_messages.py:305]: do we want to assert anything about not disconnecting, or the total ban score?

**Review by instagibbs — APPROVED** [review]
ACK a97c59f12d50d11d8859f4bbfb9fcf66de667ca0

File: test/functional/p2p_invalid_messages.py (hunk 1 above)
Comment (theStack) [reply test/functional/p2p_invalid_messages.py:291]: Nice indeed, will take it if I have to retouch.

**Comment by sr-gi** [comment]
ACK https://github.com/bitcoin/bitcoin/commit/a97c59f12d50d11d8859f4bbfb9fcf66de667ca0

File: test/functional/p2p_invalid_messages.py (hunk 1 above)
Comment (sr-gi) [reply test/functional/p2p_invalid_messages.py:291]: Agreeing with @instagibbs, the inline for loop init feels more idiomatic, but certainly not a blocker

**Comment by achow101** [comment]
ACK a97c59f12d50d11d8859f4bbfb9fcf66de667ca0

**Pull request merged**